### DISCUSSION TITLE
Run github actions in Dagger?

### DISCUSSION THREAD

**Discussion by marcosnils** [dagger/dagger#3358] [Ideas]
This topic came up in 10/13 community call. The main reason to do this is because some actions provide really nice features in Github like adding line, pr or issue comments. For example, the golangci-lint action (https://github.com/golangci/golangci-lint-action#golangci-lint-action) can add per-line feedback as how in the README.

Additionally, the act project (https://github.com/nektos/act) is already doing something to provide some workarounds to this.

There's a bunch of quirks on how it'd work but the main concerns are: 

- Can we support both container and typescript based action?
- What do we do when users run those steps in a different non-GA environment?

cc @shykes @aluzzardi @vito

**Comment by aluzzardi** [comment]
To add from experience from 0.2 and 0.3 CI (i.e. dogfooding): Typically what I end up doing is:

- dagger plan has all the lint targets (can be invoked individually or all together): e.g. go, markdown, ...
- locally, I (and everyone else using the project) invokes them all at once (e.g. `mage lint:all`, `dagger do lint`, ...)
- however, on CI, we selectively run some of the linters on Dagger (e.g. `dagger do lint cue`, `mage lint:codegen`) and use GitHub actions in parallel when available (e.g.  [golangci-lint action](https://github.com/golangci/golangci-lint-action#golangci-lint-action) )

This is because the `golangci` action adds per-line feedback directly in the PR, among other GitHub-only stuff. Using `golangci` directly would be a downgrade in experience.

I could in theory DIY (call `golangci` docker image, parse the result, call GH API to add annotations) but it feels like reinventing the wheel since this is exactly what the action is doing.

So in the end I end up with a hybrid configuration where CI and local don't match (e.g. added the markdown-lint GH action but forgot to add that to the plan which is used locally, etc).

If I could run the `golangci-lint` action directly in the plan, then local and CI would be exactly the same. Maybe it's wrapped in a `golangci-lint` extension that selectively calls the docker image or the action depending on where it's running.

Beside having the environments match, another advantage is now the action itself is cached. On `main` (0.2) we have a bunch of GitHub workflow file patterns to trigger depending on what changed (e.g. run `golangci-lint` only if `.go` files have changed).

By running the action in dagger itself, I could just rely on the cache to not invoke the action again if there's no need (both locally and on CI). Also could include file patterns directly in the plan (e.g. `include: []string{"*.go"}`)

**Comment by sheldonhull** [comment]
A great example of this is [trunk](https://github.com/trunk-io/trunk-action).
I'm using it right now to simplify all my ci checks. In some ways it's similar to the goals of dagger in providing a single unified way to stop dealing with environment issues and instead run an entire toolchain with a single command. 


For me that means `trunk install` in a repo and then just `trunk check --ci` to validate a pull request. When pushing, it adds annotations on failures. It's saved me a lot of time in setting up tooling.
